src/parser/TripleComponent.h

```diff
@@ -215,54 +215,73 @@ class TripleComponent {
+      return *id;
     }
     return std::nullopt;
   }
 
   // Same as the above, but also consider the given local vocabulary. If the
-  // string is neither in `vocabulary` nor in `localVocab`, it will be added to
-  // `localVocab`. Therefore, we get a valid `Id` in any case. The modifier is
-  // `&&` because in our uses of this method, the `TripleComponent` object is
-  // created solely to call this method and we want to avoid copying the
-  // `std::string` when passing it to the local vocabulary.
+  // Literal or IRI is neither in `vocabulary` nor in `localVocab`, it will be
+  // added to `localVocab`. Therefore, we get a valid `Id` in any case. The
+  // modifier is `&&` because in our uses of this method, the `TripleComponent`
+  // object is created solely to call this method and we want to avoid copying
+  // the Literal or IRI when passing it to the local vocabulary.
   template <typename Vocabulary>
   [[nodiscard]] Id toValueId(const Vocabulary& vocabulary,
                              LocalVocab& localVocab,
                              const EncodedIriManager& encodedIriManager) && {
-    std::optional<Id> id = toValueId(vocabulary, encodedIriManager);
-    if (!id) {
-      // If `toValueId` could not convert to `Id`, we have a string, which we
-      // look up in (and potentially add to) our local vocabulary.
-      AD_CORRECTNESS_CHECK(isLiteral() || isIri());
-      using LiteralOrIri = ad_utility::triple_component::LiteralOrIri;
-      auto moveWord = [&]() -> LiteralOrIri {
-        if (isLiteral()) {
-          return LiteralOrIri{std::move(getLiteral())};
-        } else {
-          return LiteralOrIri{std::move(getIri())};
-        }
-      };
-      // NOTE: There is a `&&` version of `getIndexAndAddIfNotContained`.
-      // Otherwise, `newWord` would be copied here despite the `std::move`.
-      id = Id::makeFromLocalVocabIndex(
-          localVocab.getIndexAndAddIfNotContained(moveWord()));
+    auto idOrBounds = toValueIdOrBounds(vocabulary, encodedIriManager);
+    if (auto* id = std::get_if<Id>(&idOrBounds)) {
+      return *id;
     }
-    return id.value();
+    using Bounds = std::pair<VocabIndex, VocabIndex>;
+    AD_CORRECTNESS_CHECK(std::holds_alternative<Bounds>(idOrBounds));
+    auto [lower, upper] = std::get<Bounds>(idOrBounds);
+    // If `toValueId` could not convert to `Id`, we have a Literal or Iri,
+    // which we look up in (and potentially add to) our local vocabulary.
+    AD_CORRECTNESS_CHECK(isLiteral() || isIri());
+    using LiteralOrIri = ad_utility::triple_component::LiteralOrIri;
+    auto moveWord = [&]() -> LiteralOrIri {
+      if (isLiteral()) {
+        return LiteralOrIri{std::move(getLiteral())};
+      } else {
+        return LiteralOrIri{std::move(getIri())};
+      }
+    };
+    // NOTE: There is a `&&` version of `getIndexAndAddIfNotContained`.
+    // Otherwise, `newWord` would be copied here despite the `std::move`.
```

Comment:
This comment is confusing, there is no `std::move` here and just a `moveWord`, not a `newWord`.